%%BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
\usepackage{algorithmic}
\usepackage{rotating}
\usepackage{amsmath}
\usepackage{multirow}
\usepackage{enumerate}
\usepackage{algorithm}
\usepackage{subfigure}
\usepackage{graphicx}
\usepackage{adjustbox}
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\def\includegraphic{}
%\def\includegraphics{}

%%% Put your definitions there:
\startlocaldefs

\newcommand{\alexei}[1]{{\textcolor{cyan} {Alexei says: #1} }}


\endlocaldefs


\begin{document}
\begin{frontmatter}
\begin{fmbox}
\dochead{Methodology article}

\title{Improving the performance of Bayesian phylogenetic inference under relaxed clock models}

\author[
   addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
   email={rzha419@aucklanduni.ac.nz}
]{\inits{RZ}\fnm{Rong} \snm{Zhang}}
\author[
   addressref={aff1,aff2},
   corref={aff1,aff2},                       % id of corresponding address, if any
   email={alexei@cs.auckland.ac.nz}   % email address
   %noteref={n1},                        % id's of article notes, if any
]{\inits{AJD}\fnm{Alexei} \snm{Drummond}}

\address[id=aff1]{%                           % unique id
  \orgname{School of Computer Science, University of Auckland}, % university, etc
  \city{Auckland},                              % city
  \cny{New Zealand}                                    % country
}

\address[id=aff2]{%                           % unique id
  \orgname{School of Biological Sciences, University of Auckland}, % university, etc
  \city{Auckland},                              % city
  \cny{New Zealand}                                    % country
}

%\address[id=aff2]{%
%  \orgname{},
%  \street{},
%  \postcode{}
%  \city{},
%  \cny{}
%}

%\begin{artnotes}
%%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
%\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{abstractbox}
\begin{abstract}
\parttitle{Background}
Bayesian MCMC has become a common approach for phylogenetic inference. But the growing size of molecular sequence data sets has created a pressing need to improve the computational efficiency of Bayesian phylogenetic inference algorithms.
%Phylogenetic trees describe the evolutionary history of a group of species. With the rapid accumulation of both genetic and morphological data, more and more models have been developed to infer phylogenetic trees. In recent years, Bayesian MCMC has become a common approach for phylogenetic inference. However, the huge amount of available data challenge the efficiency of phylogenetic models.
\parttitle{Results}
This paper develops a new algorithm to improve the efficiency of Bayesian phylogenetic inference for models that include a per-branch rate parameter. In a Markov chain Monte Carlo algorithm, the presented proposal kernel changes evolutionary rates and divergence times at the same time, under the constraint that the implied genetic distances remain constant. Specifically, the proposal operates on the divergence time of an internal node and the three adjacent branch rates. For the root of a phylogenetic tree, there are three strategies discussed, named Simple Distance, Small Pulley and Big Pulley. Note that Big Pulley is able to change the tree topology, which enables the operator to sample all the possible rooted trees consistent with the implied unrooted tree. To validate its effectiveness, a series of experiments have been performed by implementing the proposed operator in the BEAST2 software.
\parttitle{Conclusions}
The results demonstrate that the proposed operator is able to improve the performance by giving better estimates for a given chain length and by using less running time for a given level of accuracy. Measured by effective samples per hour, use of the proposed operator results in overall mixing that is up to an order of magnitude more efficient than the current operators in BEAST2 on both real and simulated data sets.
\end{abstract}

\begin{keyword}
\kwd{Bayesian MCMC}
\kwd{Bayesian phylogenetics}
\kwd{Proposal kernel}
\kwd{Genetic distances}
\kwd{Divergence times}
\kwd{Evolutionary rates}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout
\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%
\section*{Background}
Bayesian phylogenetics puts an emphasis on estimating a probability distribution over parameters of interest, including the phylogenetic tree topology and divergence times, given the data. The Metropolis-Hastings Markov chain Monte Carlo (MCMC) \cite{metropolis1953equation,hastings1970monte} algorithm has been the primary computational tool used in Bayesian phylogenetics for sampling from the posterior distribution. This paper is aimed at improving the performance of the relaxed clock model in Bayesian phylogenetic analysis.

Early implementations of Bayesian phylogenetic inference assumed a strict molecular clock where the evolutionary rates are the same at every branch \cite{zuckerkandl1965evolutionary}. This was the preferred method for estimating divergence times \cite{yang1997bayesian,rannala2003bayes}. The introduction of relaxed molecular clocks allowed for the estimation of divergence times \cite{thorne1998estimating} and phylogeny \cite{drummond2006relaxed} in the presence of rate heterogeneity among branches. Since then, the relaxed clock model has been widely applied, such as the study of Nothofagus \cite{knapp2005relaxed} and flowering plants \cite{smith2010uncorrelated}. 
%More specifically, as branch rates vary throughout a phylogenetic tree, it is able to identify regional gene bias in cooperation with fossil calibrations, according to Susanne's research in plant dispersal events \cite{renner2005relaxed}. 
Many aspects of the performance and accuracy of relaxed clock models have subsequently been investigated (e.g. \cite{ho2005accuracy}, \cite{lepage2007general}).

Bayesian phylogenetic inference via MCMC is computationally intensive for large data sets.
Two approaches to improve efficiency are (i) by making faster likelihood calculations, and (ii) by incorporating more effective proposal kernels. Calculating the phylogenetic likelihood is computationally expensive.
Hence, researchers have tried many ways to tackle the computation burden in the likelihood calculations, such as detection of repeating sites \cite{kobert2017efficient},  approximate methods (e.g. \cite{guindon2010bayesian,reis2011approximate}) and the use of parallelisation strategies (e.g. BEAGLE \cite{ayres2011beagle}).

However the overall efficiency of the sampling process also depends strongly on the construction of the proposal mechanism. An effective proposal mechanism is proficient at exploring the posterior distribution, and can do so with fewer steps in the MCMC chain. Therefore fewer likelihood calculations are required, since each step in the chain that changes the tree or substitution parameters requires a likelihood calculation.
%More efficient proposal mechanisms require fewer steps in the MCMC chain and thus fewer likelihood computations, since each step of the MCMC chain entails a likelihood calculation.

A major limitation in Bayesian MCMC analysis of phylogeny lies in the efficiency with which operators sample the tree space \cite{lakner2008efficiency,hohna2012guided}. Fast and reliable estimation is dependent on a good mixture of operators, since the posterior distribution often exhibits correlations between the tree and other random variables.
%Many such tree proposal mechanisms have been described \cite{}. %According to the work in Ref.\cite{hohna2011guided}, the two common operators, i.e. prune-and-regraft and subtree-swap, both contribute to a tree with low likelihood since they propose a new tree by random movement of the current tree. So the authors introduced two new operators by proposing a state from a discrete set of possible proposals and narrowing the proposal distribution to the more likely proposals. Their experimental results proved that the two operators have faster average run time and more accurate predictability.

In this paper, we present a novel operator that works alongside standard operators by proposing moves within a subspace of constant genetic distances.
Namely, the proposed operator changes both divergence times of nodes and neighbouring branch rates so that the implied genetic distances are not changed. For time-reversible substitution models the phylogenetic likelihood will also be unchanged under this operation. The proposed operator has been implemented and tested in BEAST2 \cite{bouckaert2014beast}.

\section*{Preliminaries}
\subsection*{Bayesian MCMC}
Let $\bf{D}$, $g$ and $\bf{\Phi}$ denote the data, phylogenetic time-tree and a set of evolutionary parameters respectively.
The time-tree $g = \{E,\bf{t}\}$ consists of a directed edge graph, $E$, defining a rooted tree topology on a set of labelled taxa and a set of associated divergence times $\bf{t}$ (for details see e.g. \cite{drummond2002estimatingdata}).
 The posterior probability density can be calculated using equation \ref{bayes}. It consists of prior distributions for the  tree and the parameters, a phylogenetic likelihood that conveys information from data, and the posterior distribution to be inferred. These are denoted in the form of probability densities by $p(g) $, $p(\bf{\Phi})$, $p({\bf{D}}|g,\bf{\Phi} )$, $p(g,{\bf{\Phi}}|{\bf{D}})$ respectively. From a Bayesian perspective, the phylogenetic trees and the parameters are random variables described by a posterior probability distribution given the observed data ${\bf{D}}$.
%As a result, it is possible to jointly infer evolutionary history of species  by incorporating various source of information such as molecular data and fossil records.
\begin{equation}\label{bayes}
p(g,{\bf{\Phi}} |{\bf{D}}) = \frac{{p ({\bf{D}}|g,{\bf{\Phi}} ) p (g) p ({\bf{\Phi}} )}}{{p (\bf{D})}}
\end{equation}

However, due to the state space being high dimensional and the marginal likelihood being infeasible to calculate, MCMC is adopted to sample the posterior distribution. Specifically, MCMC algorithms construct a Markov chain whose stationary distribution is the posterior distribution $p(g,{\bf{\Phi}} |{\bf{D}})$, in such a way that the computation of the marginal likelihood $p ({\bf{D}})$ is avoided.
%In this paper, a Metropolis-Hastings algorithm (MH-MCMC) is implemented, which generates a new state given the current state through a proposal density and accepts the new state with a prescribed probability \cite{metropolis1953equation,hastings1970monte}.
\subsection*{Tree proposals}
\label{treeproposals}
We use the term ``operator" to describe an algorithm that can be used to draw a new state $\theta'$ given an existing state $\theta = \{g,{\bf{\Phi}}\}$ from a specific proposal kernel $q(\theta'|\theta)$ and also return the Hastings-Green ratio for the proposed state transition \cite{hastings1970monte,green1995reversible}.
%In MH-MCMC, an operator implements the proposal density and provided relevant information to calculate the acceptance probability for the next proposed state.

Standard na\"{i}ve operators such as the random walk operator propose the new state $\theta^\prime$ by adding a random variate to a component of the current state $\theta$ \cite{suchard2005stochastic}. Similarly, scale operators multiply a subset of the current state by a random scale factor \cite{higuchi1997monte}. They are suitable for working on a single random variable, or a single component of the model, for example the population size parameter of the coalescent tree prior. Standard operators for the tree topology and divergence times include the subtree slide operator, Wilson-balding and narrow exchange operators \cite{drummond2002estimatingdata,hohna2008clock}.
%On top of this, there are also some extended operators available to help give the better performance. For instance, the prune-and-regraft operator selects a random subtree and reattaches the subtree at a new random branch \cite{hohna2011guided}.

In this paper, the novelty of the proposed operators lies in maintaining the genetic distance $d$ while changing the rate $r$ and divergence time $t$. The reason is that the likelihood along one branch is constant if its distance is fixed, i.e. $d = r \times t$, noting that the likelihood is calculated based on transition probability matrix for each branch of ${e^{\textbf{Q}d_i}}$, where $d_i$ is the branch length in units of substitutions per site for branch $i$. In this way, the joint distribution on rates and divergence times can be explored without proposing states that would adversely affect the phylogenetic likelihood.

\subsection*{Uncorrelated relaxed clock model}
Molecular clocks model how molecular sequences evolve along branches in the phylogenetic tree, so that a time tree can be reconciled with the genetic distances between sequences. In this paper, uncorrelated relaxed clock models are adopted, where the rates are drawn independently and identically from a given prior distribution, such as the log-normal distribution \cite{drummond2006relaxed}. As a result, the rates can vary markedly between parent and child branches.

Referring to the Bayesian framework in equation (\ref{bayes}), the joint inference of evolutionary rates ${\bf{r}}$ and the time tree $g$ can be obtained by the conditional distribution in equation \ref{bayes1}:
\begin{equation}\label{bayes1}
p(g,{\bf{r}},{\bf{\Phi}} |{\bf{D}}) = \frac{{p({\bf{D}}|g,{\bf{r}},{\bf{\Phi}} )p({\bf{r}})p(g)p({\bf{\Phi}} )}}{{p({\bf{D}})}} \text{,}
\end{equation}
where $p({\bf{r}})$ is the prior for rates specified in uncorrelated relaxed clock model. In the constructed Markov chain, the operator proposes a new state ${\theta^\prime} = ({\bf{r}^\prime},g^\prime,{\bf{\Phi}^\prime})$, from the original state $\theta =  \{ {\bf{r}},g,{\bf{\Phi}}\}$.

While the proposed operator is introduced based on uncorrelated clock models, it could equally be applied to any other relaxed clock that applies a rate parameter to each branch, such as autocorrelated clock models \cite{thorne1998estimating}.

\section*{Methods}
In this section, we define the Constant Distance Operator. Figure \ref{flowchart} illustrates the flow chart of the proposed operators. In a phylogenetic tree, the node to operate on is denoted by \textbf{X} and the Constant Distance Operator works differently on internal nodes and the root node. The details of the operations are introduced step by step in the following subsections.
\subsection*{Operations on internal nodes}
Figure \ref{internalnodes} represents the tree (or subtree) with the node \textbf{X} that is randomly selected from among the internal nodes. Let $g$ be the tree in the current state. The following steps propose a new divergence time in ${g}'$ and three rates in $\bf{r}'$.

\emph{Step 1} Identify the parent node and two child nodes of \textbf{X}, denoted by \textbf{P}, \textbf{L} and \textbf{R} respectively.

\emph{Step 2} Denote the nodes times of \textbf{X}, \textbf{P}, \textbf{L} and \textbf{R} by $t_X$, $t_P$, $t_L$, $t_R$ respectively. Denote the rates on the branches above the nodes by $r_X$, $r_L$ and $r_R$ respectively.

\emph{Step 3} Propose a new node time for \textbf{X} by ${t_X}' \leftarrow {t_X} + a$, where $a$ follows a Uniform distribution with a symmetric window size $w$, i.e. $a \sim$ Uniform[-w, +w], for some window size $w$. Make sure that the proposed time is valid, i.e. $\max \{ {t_L},{t_R}\}  < {t_X}' < {t_P}$ holds. Otherwise, we reject the proposal.

\emph{Step 4} Propose new rates by using equation \ref{int_rate}.
\begin{equation}
 \label{int_rate}
{r_X}' = \frac{{r_X} \times ({{t_P} - {t_X}})}{{{t_P} - {t_X}'}}\\
{r_L}' = \frac{{{r_L} \times ({{t_X} - {t_L}})}}{{{t_X}' - {t_L}}}\\
{r_R}' = \frac{{{r_R} \times ({{t_X} - {t_R}})}}{{{t_X}' - {t_R}}}
 \end{equation}

\emph{Step 5} Return the Green ratio ${\alpha_{IN}}$ (Refer to \textit{Calculating the Green Ratio} in the following subsection).

\subsection*{Operations on the root}
We present three strategies for proposing the new rates and a new divergence time for the special case when \textbf{X} is the root node. i) The Simple Distance operator only proposes a new root time. ii) Small Pulley adjusts the distances of branches on both sides of the root. iii) Big Pulley proposes a new tree topology by rearranging the root, without perturbing the unrooted tree. As is illustrated in Figure \ref{simpledistance}(a), all the operations on the root, including Big Pulley that changes the tree topology, do not change the underlying unrooted tree. For instance, no matter where the root $X$ is (either on branch $EF$ or $AE$), the operators maintain the distances ($d_{AB}$, $d_{AC}$, $d_{AD}$, $d_{BC}$, $d_{BD}$, $d_{CD}$) and preserve the unrooted tree at the same time.

\subsubsection*{Simple Distance}
Figure \ref{simpledistance} (b), (c) and (d) show the trees that are rooted at the node \textbf{X}. The original tree $g$ in the current state is shown in Figure \ref{simpledistance}(b). Similar to the operations on internal nodes, we will use the following steps to propose a new root time in ${g}'$ and two rates in $\bf{r}'$, as is illustrated in Figure \ref{simpledistance}(c). At the same time, the genetic distances of two branches linked to the root, i.e. $d_L$ and $d_R$, are kept constant constant.

\emph{Step 1} Identify the child nodes of the root \textbf{X}, denoted by \textbf{L} and \textbf{R}. Their corresponding node times and branch rates are $t_X$, $t_L$, $t_R$ and $r_L$, $r_R$.

\emph{Step 2} Propose a new node time for the root \textbf{X} by ${t_X}' \leftarrow {t_X} + a$, where $a \sim$ Uniform[-w, +w]. Make sure that ${t_X}' > \max \{ {t_L}, {t_R}\} $ holds. Otherwise, we reject the proposal.

\emph{Step 3} Propose new rates for branches on both sides of the root by using equation \ref{sim_rate}.
\begin{equation}
\label{sim_rate}
{r_L}' = \frac{{{r_L} \times ({{t_X} - {t_L}})}}{{{t_X}' - {t_L}}}\\
{r_R}' = \frac{{{r_R} \times ({{t_X} - {t_R}})}}{{{t_X}' - {t_R}}}
 \end{equation}

\emph{Step 4} Return the Green ratio ${\alpha_{SD}}$.
\subsubsection*{Small Pulley}
In contrast to Simple Distance, Small Pulley changes genetic distances of branches on both sides of the root. As is illustrated in Figure \ref{simpledistance}(d), two new rates in $\bf{r}'$ are proposed based on those in the original tree $g$. In order to maintain the total genetic distance $d_L + d_R$ of the two branches linked to the root, after ${d_L}'$ is proposed, $d_R$ will be adjusted simultaneously. In other words, Small Pulley keeps $D = d_L + d_R$ constant. The detailed process includes the following 4 steps.

\emph{Step 1} Identify the child nodes of the root \textbf{X}, denoted by \textbf{L} and \textbf{R}. Their corresponding node times and branch rates are $t_X$, $t_L$, $t_R$ and $r_L$, $r_R$. The implied genetic distances of the two branches linked to the root can be calculated by:
\begin{equation}
\label{sma1_dis}
{d_L} = {r_L} \times ({{t_X} - {t_L}})\\
{d_R} = {r_R} \times ({{t_X} - {t_R}})
 \end{equation}

\emph{Step 2} Propose a new genetic distance for $d_L$ by adding a random number that follows a Uniform distribution, i.e.  ${d_L}' \leftarrow {d_L} + b$, where $b \sim$ Uniform[-v, +v], for some window size $v$. Make sure that $0 < {d_L}' < D$ holds. Otherwise, we reject the proposal.

\emph{Step 3} Propose new rates for branches on each side of the root:
\begin{equation}
\label{sma1_rate}
{r_L}' = \frac{{{d_L}'}}{{{t_X} - {t_L}}}\\
{r_R}' = \frac{{D - {d_L}'}}{{{t_X} - {t_R}}}
 \end{equation}

\emph{Step 4} Return the Green ratio ${\alpha_{SP}}$.
\subsubsection*{Big Pulley}
Big Pulley resamples the rates and times while maintaining the implied unrooted tree in units of genetic distance. So the genetic distances between the taxa are held constant, but the location of the root in the time tree is readjusted.

Before describing the detailed steps, we introduce a method \textit{Exchange} that proposes a new root position. In Figure \ref{exchangemethod}, let (i) \textbf{X} denote the root of tree $g$, (ii) \textbf{C} and \textbf{N} denote the two child nodes of \textbf{X}, (iii) \textbf{S} and \textbf{M} denote the two child nodes of \textbf{C}. The \textit{Exchange(\textbf{M}, \textbf{N})} method involves the following steps:
\begin{itemize}
\item Swap the two nodes by pruning and regrafting, i.e. cutting \textbf{M} (\textbf{N}) at its original position and attaching it to the original position of \textbf{N} (\textbf{M}).
\item Propose ${d_C}' \leftarrow {d_C} + b$, where $b \sim$ Uniform[-v, +v]. Make sure that $0 < {d_C}' < D$ holds, where $D = {d_C} + {d_{N}}$. Otherwise, we reject the proposal.
\item The distances on the other three branches, i.e. $d_S$, $d_{M}$ and $d_{N}$, will be adjusted:
\begin{equation}\label{big_dis}
{d_S}' = {d_S}\\
{d_{M}}' = {d_{M}} - {d_C}'\\
{d_{N}}' = {d_{N}} + {d_C}
\end{equation}
\end{itemize}

As can be seen from the above descriptions, the method \textit{Exchange(\textbf{M}, \textbf{N})} swaps two nodes and adjusts distances ($d_S$, $d_{M}$, $d_{N}$ and $d_{C}$) on the four branches so as to maintain the implied genetic distances among three taxa \textbf{S}, \textbf{M} and \textbf{N}.

Additionally, operations in Big Pulley vary depending on the shape of phylogenetic tree. In Figure \ref{treeshape}, a symmetric tree is shown on the left, in which both the child nodes of the root have child nodes, i.e. \textbf{L} having children \textbf{H1}, \textbf{H2} and \textbf{R} having children \textbf{H3}, \textbf{H4}. But in the asymmetric tree on the right, only one of the child nodes of the root has child nodes below it, i.e. \textbf{O} having children \textbf{G1}, \textbf{G2}. But the other child node \textbf{Y} doesn't have any offsprings, which may be a tip or a sampled ancestor. The corresponding operations are detailed in the following two parts.
\paragraph*{Symmetric tree}

For the symmetric tree in Figure \ref{treeshape}, the operations are illustrated in Figure \ref{symmetric}, after which one of the four possible trees (\textcircled1 \textcircled2 \textcircled3 \textcircled4) will be proposed. The detailed process is described in Algorithm \ref{symmetric tree proposal}.

\begin{algorithm}
\caption{Proposal for symmetric trees in Big pulley}
\label{symmetric tree proposal}
\begin{algorithmic}
\STATE \COMMENT{Step 1: Identify the child nodes of the root \textbf{X}, denoted by \textbf{L} and \textbf{R}. Correspondingly, the node times are denoted by $t_X$, $t_L$, $t_R$. And the child nodes below them are denoted by \textbf{H1}, \textbf{H2}, \textbf{H3} and \textbf{H4}.}
\STATE Let \textbf{X} be the root of the tree.
\STATE Let \textbf{L} and \textbf{R} be the left child and right child of \textbf{X}, respectively.
\STATE \COMMENT{Step 2: Propose a new node time for the root \textbf{X}.}
\STATE $a \sim$ Uniform[-w, +w]
\STATE ${t_X}' \leftarrow {t_X} + a$
\STATE \COMMENT{Step 3: Propose a new node time either for \textbf{L} or \textbf{R}, and adjust adjacent rates.}
%\STATE Draw a random number $\sigma_1$ from Uniform(0,1).
\IF{${\sigma_1} \sim Uniform(0,1) < 0.5$}
\STATE Pick \textbf{L} and propose a new node time by ${t_L}'  \leftarrow {t_L} + {a_1}$, where ${a_1} \sim$ Uniform[-w, +w]. 
\IF{${t_R} < {t_L}' < {t_X}'$}
%\STATE Draw a random number $\sigma_2$ from Uniform(0,1). 
\IF{${\sigma_2} \sim Uniform(0,1) < 0.5$}
\STATE Apply \textit{Exchange (\textbf{H1}, \textbf{R})} and propose tree \textcircled1.
\ELSE
\STATE Apply \textit{Exchange (\textbf{H2}, \textbf{R})} and propose tree \textcircled2.
\ENDIF
\ELSE
\STATE Reject the proposal.
\ENDIF
\ELSE
\STATE Pick \textbf{R} and propose a new node time by ${t_R}'  \leftarrow {t_R} + {a_2}$, where ${a_2} \sim$ Uniform[-w, +w]. 
\IF{${t_L} < {t_R}' < {t_X}'$}
%\STATE Draw a random number $\sigma_3$ from Uniform(0,1). 
\IF{${\sigma_3} \sim Uniform(0,1) < 0.5$}
\STATE Apply \textit{Exchange (\textbf{H3}, \textbf{L})} and propose tree \textcircled3.
\ELSE
\STATE Apply \textit{Exchange (\textbf{H4}, \textbf{L})} and propose tree \textcircled4.
\ENDIF
\ELSE
\STATE Reject the proposal.
\ENDIF
\ENDIF
\STATE \COMMENT{Step 4: Update the rates on the corresponding branches.} 
\STATE \COMMENT{Step 5:  Return the Green ratio ${\alpha_{BP}}$.}
\end{algorithmic}
\end{algorithm}

For example, suppose we are going to propose tree \textcircled1. After the new node times for the root \textbf{X} and \textbf{L} are proposed, we apply the method by \textit{Exchange (\textbf{H1}, \textbf{R})}, so that four distances are adjusted, as follows:

\begin{equation}\label{example1.1}
{d_{H1}}' = {d_{H1}} - {d_L}'  \\
{d_{H2}}' = {d_{H2}} \\
{d_L}' = {d_L} + b \\
{d_R}' = {d_L} + {d_R}
\end{equation}

Finally, in this example the new rates would be updated by:

\begin{equation}\label{example1.2}
{r_{H1}}' = \frac{{{d_{H1}}'}}{{{t_X}' - {t_{H1}}}} \\
{r_{H2}}' = \frac{{{d_{H2}}'}}{{{t_L}' - {t_{H2}}}} \\
{r_L}' = \frac{{{d_L}'}}{{{t_X}' - {t_L}'}} \\
{r_R}' = \frac{{{d_R}'}}{{{t_L}' - {t_R}}}
\end{equation}

\paragraph*{Asymmetric tree}

For an asymmetric tree such as in Figure \ref{treeshape} we would operate as illustrated in Figure \ref{asymmetric}, in which there are three possible trees (\textcircled5 \textcircled6 \textcircled7). The operations are detailed in Algorithm \ref{asymmetric tree proposal}.

\begin{algorithm}
\caption{Proposal for asymmetric trees in Big pulley}
\label{asymmetric tree proposal}
\begin{algorithmic}
\STATE \COMMENT{Step 1: Identify the extant child node of the root \textbf{X}, which has two child nodes below and is denoted by \textbf{O}. The extinct child node of the root, which does not have any child nodes, is denoted by \textbf{Y}. The node times of the root \textbf{X},  \textbf{Y}, \textbf{O} and its child nodes are denoted by ${t_X}$, ${t_Y}$, ${t_O}$, ${t_{G1}}$ and ${t_{G2}}$ respectively.}
\STATE Let \textbf{X} be the root of the tree.
\STATE Let \textbf{O} be the child of \textbf{X} that has children, and let \textbf{Y} be the child of \textbf{X} that does not have children. 
\STATE \COMMENT{Step 2: Propose a new node time for the root \textbf{X}.}
\STATE $a \sim$ Uniform[-w, +w]
\STATE ${t_X}' \leftarrow {t_X} + a$
\STATE \COMMENT{Step 3: Propose a new node time for the node \textbf{O}.}
\STATE ${a_3} \sim$ Uniform[-w, +w]
\STATE ${t_O}' \leftarrow {t_O} + {a_3}$
\IF{${t_O}'<{t_Y}  $  or ${t_O}'> {t_X}'$}
\STATE Reject the proposal.
\ENDIF
\STATE \COMMENT{Step 4: Adjust the distances according to the tree corresponding topologies.}
\IF{${t_O}' > \max \{ {t_{G1}},{t_{G2}}\} $ or ${t_{G1}} = {t_{G2}}$}
%\STATE Draw a random number $\sigma_4$ from Uniform(0,1). 
\IF{${\sigma_4} \sim Uniform(0,1) < 0.5$}
\STATE Apply \textit{Exchange (\textbf{G1}, \textbf{Y})} and propose tree \textcircled5.
\ELSE
\STATE Apply \textit{Exchange (\textbf{G2}, \textbf{Y})} and propose tree \textcircled6.
\ENDIF
\ELSIF {$\min \{ {t_{G1}},{t_{G2}}\}  < {t_O}' < \max \{ {t_{G1}},{t_{G2}}\} $}
\STATE Exchange the older child of \textbf{O} and \textbf{Y}.  (For the asymmetric tree in Figure \ref{treeshape}, we apply \textit{Exchange (\textbf{G1}, \textbf{Y})} and propose tree \textcircled7).
\ELSIF{${t_O}' < \min \{ {t_{G1}},{t_{G2}}\}$}
\STATE Reject the proposal.
\ENDIF
\STATE \COMMENT{Step 4: Update the rates on the corresponding branches.}
\STATE \COMMENT{Step 5:  Return the Green ratio ${\alpha_{BP}}$.}
\end{algorithmic}
\end{algorithm}

To give an example, assume we are going to propose tree \textcircled5. Firstly, ${t_X}'$ and ${t_O}'$ are proposed in \emph{Step 3} and \emph{Step 4}. Then, in \emph{Step 4}, the method \textit{Exchange (\textbf{G1}, \textbf{Y})} is applied, after which the four distances are adjusted as follows:

\begin{equation}\label{example2.1}
{d_{G1}}' = {d_{G1}} - {d_O}'  \\
{d_{G2}}' = {d_{G2}}  \\
{d_O}' = {d_O} + b  \\
{d_Y}' = {d_Y} + {d_O}
\end{equation}

And the four rates are updated as follows:

\begin{equation}\label{example2.2}
{r_{G1}}' = \frac{{{d_{G1}}'}}{{{t_X}' - {t_{G1}}}} \\
{r_{G2}}' = \frac{{{d_{G2}}'}}{{{t_O}' - {t_{G2}}}} \\
{r_O}' = \frac{{{d_O}'}}{{{t_X}' - {t_O}'}} \\
{r_Y}' = \frac{{{d_Y}'}}{{{t_O}' - {t_Y}}}
\end{equation}

\subsection*{Calculating the Green ratio}
MCMC operators must use reversible proposal distributions to satisfy the detailed balance requirements of the MCMC algorithm (Refer to  Appendix section 1 for more details). Therefore, all four of our operators involve a final step of calculating the Green ratio for the proposal.

According to the third and fourth steps in the operations for internal nodes, three rates on the branches linked to the selected internal node are proposed by one random number $a$ that is used to change the node time. There are four parameters involved in this proposal, comprised of a 3-dimensional rate space and a 1-dimensional time space. The proposed operator utilises one random number in time space and makes changes in both time and rate space, which leads to a dimension-matching problem. To solve this dimension-matching problem, as is mentioned in Green's paper \cite{green1995reversible}, it is necessary to construct the Jacobian matrix.  In equation (\ref{JacobianMatrix}), ${\mathbf{J_1}}$ deals with the parametric spaces before the proposal in vector ${\mathbf{IN}} = [{t_X},{r_X},{r_L},{r_R}]$ and after the proposal in vector ${\mathbf{OUT}} = [{t_X}',{r_X}',{r_L}',{r_R}']$.
\begin{equation}\label{JacobianMatrix}
{\mathbf{J_1}} = \left[ {\begin{array}{*{20}{c}}
  {\frac{{\partial {\mathbf{f}}}}{{\partial {t_X}}}}&{\frac{{\partial {\mathbf{f}}}}{{\partial {r_X}}}}&{\frac{{\partial {\mathbf{f}}}}{{\partial {r_L}}}}&{\frac{{\partial {\mathbf{f}}}}{{\partial {r_R}}}}
\end{array}} \right] = \left[ {\begin{array}{*{20}{c}}
  {\frac{{\partial {f_1}}}{{\partial {t_X}}}}&{\frac{{\partial {f_1}}}{{\partial {r_X}}}}&{\frac{{\partial {f_1}}}{{\partial {r_L}}}}&{\frac{{\partial {f_1}}}{{\partial {r_R}}}} \\
  {\frac{{\partial {f_2}}}{{\partial {t_X}}}}&{\frac{{\partial {f_2}}}{{\partial {r_X}}}}&{\frac{{\partial {f_2}}}{{\partial {r_L}}}}&{\frac{{\partial {f_2}}}{{\partial {r_R}}}} \\
  {\frac{{\partial {f_3}}}{{\partial {t_X}}}}&{\frac{{\partial {f_3}}}{{\partial {r_X}}}}&{\frac{{\partial {f_3}}}{{\partial {r_L}}}}&{\frac{{\partial {f_3}}}{{\partial {r_R}}}} \\
  {\frac{{\partial {f_4}}}{{\partial {t_X}}}}&{\frac{{\partial {f_4}}}{{\partial {r_X}}}}&{\frac{{\partial {f_4}}}{{\partial {r_L}}}}&{\frac{{\partial {f_4}}}{{\partial {r_R}}}}
\end{array}} \right]\text{,}
\end{equation}
where the functions ${f_1}$, ${f_2}$, ${f_3}$ and ${f_4}$ represent how the operator makes a proposal. After substituting equation (\ref{int_rate}) in equation (\ref{JacobianMatrix}), the Green ratio for the internal nodes can be derived:
\begin{equation}\label{HR1}
{\alpha_{IN}} = \frac{{p ( - a)}}{{p (a)}}\left| {\mathbf{J_1}} \right| = \frac{{{t_P} - {t_X}}}{{{t_P} - {t_X}'}} \times \frac{{{t_X} - {t_L}}}{{{t_X}' - {t_L}}} \times \frac{{{t_X} - {t_R}}}{{{t_X}' - {t_R}}}\text{,}
\end{equation}
where the proposal density ${p ( - a)}$ is equal to ${p ( a )}$ since the random number $a$ is drawn from Uniform distribution.

Likewise, the Green ratio for Simple Distance, Small Pulley and  Big Pulley can be obtained:
\begin{equation}\label{HR2}
{\alpha_{SD}} = \frac{{{t_X} - {t_L}}}{{{t_X}' - {t_L}}} \times \frac{{{t_X} - {t_R}}}{{{t_X}' - {t_R}}}\text{,}
\end{equation}
\begin{equation}\label{HR3}
{\alpha_{SP}}  = 1\text{,}
\end{equation}
\begin{equation}\label{HR4}
{\alpha_{BP}} = \mu \times \frac{{{t_X}' - {t_C}}}{{{t_X}' - {t_C}'}} \times \frac{{{t_C} - {t_S}}}{{{t_C}' - {t_S}}} \times \frac{{{t_C} - {t_{N1}}}}{{{t_X}' - {t_{N1}}}} \times \frac{{{t_X} - {t_{N2}}}}{{{t_C}' - {t_{N2}}}}\text{,}
\end{equation}
where $\mu = p(g', g) / p(g, g')$ is defined as the proposal ratio of topology change and is obtained by Algorithm \ref{alg1}. More details of how to calculate the determinant of the Jacobian matrix are explained in Appendix section 1.

\section*{Results}
To validate the correctness and determine the efficiency, we conducted a series of experiments by implementing the Constant Distance operator in BEAST2 \cite{bouckaert2014beast}.

First, we perform a well-calibrated simulation study, which tests our operator alongside existing operators. Correctness was further confirmed by sampling trees from the prior distribution i.e. without data (Refer to Appendix section 2 for more details). By comparing effective sample sizes (ESS) \cite{Tracer} and running times, it is demonstrated that the performance is improved when including our proposed operator. Finally, the posterior correlation of rates and node times are discussed.

\subsection*{Well-calibrated simulation study}
A well-calibrated simulation study is a powerful tool for evaluating and validating the implementation of a Bayesian model \cite{dawid1982well}.

Figure \ref{modelvalidation} shows the Bayesian model used in this study, which includes the evolutionary model and the prior distributions of parameters. As is shown in the figure, the sequence alignment is simulated by a phylogenetic continuous-time Markov chain in BEAST2. It contains a substitution rate matrix given by the HKY85 \cite{hasegawa1985dating} model and a substitution tree determined by an uncorrelated relaxed clock model and Yule model. More specifically, base frequencies $\bf{\pi}$ follow a Dirichlet distribution and the transition-transversion ratio $\kappa$ follows a log-normal prior distribution. The distribution of node times is described in a Yule tree $\psi$ with hyperparameter birth rate $\lambda $ following a log-normal distribution. The rates $r_i$ follow a log-normal distribution with mean of 1 and standard deviation $s_1$ following a hyperprior distribution.

First, we sampled parameters and trees from the full model 100 times. The random parameters included: standard deviation of rates across branches$s_1$, birth rate $\lambda $, base frequencies $\bf{\pi}$ and transition-transversion bias $\kappa$. Second, we simulated nucleotide alignments using the simulated parameters. In total, 100 data sets were simulated, each with 120 taxa. Third, we used BEAST2 with the Constant Distance operator to infer the tree and parameters from each of the 100 simulated data sets in turn. Finally, the posterior estimates of the parameters were compared with the real values that were used to simulate the corresponding sequence alignment. The comparisons are shown in Figures \ref{LargeTree}.

These results show that the true values of the parameters are within the 95\% highest posterior density (HPD) interval approximately 95\% of the time (Table \ref{num_hpd}). This well-calibrated simulation study formed part of the validation of our implementation of the Constant Distance operator.

\subsection*{Performance comparison}
To evaluate the performance of Constant Distance operator in a Bayesian phylogenetic analysis, we explored the time required to adequately sample the posterior distribution. This was achieved by examining (i) the total time taken by BEAST2 to complete the MCMC inference (running time), and (ii) the effective sample size (ESS) of the sampled parameters. The effective sample size of a parameter is the number of effectively independent samples from the posterior distribution. Larger ESS indicates a better approximation of the marginal posterior distribution of the parameter. We used Tracer \cite{Tracer} to compute ESS.

For each dataset, we compared two operator configurations. 1) Using the current operators in BEAST2 to sample discrete rate categories (Category). 2) Using the Constant Distance operator to sample continuous rates specified by an uncorrelated related clock model (Cons). The Category configuration is the default setting in BEAST 2.5. We aim to compare the performance of the Constant Distance operator to that of the existing operator schedule. In each configuration, the data set was analyzed 20 times with the prior distributions and all other model specifications held constant. The details of operator weights used are given in Appendix 3.1.

We performed the analysis on two sets of simulated sequence alignment (See Appendix 3.2 for more details). The simulated data sets both have 20 taxa but different sequence lengths, i.e. one data set containing 500 sites, the other containing 1,000 sites. Moreover, we used four real data sets to further evaluate the performance of Constant Distance operator, including a primate data set \cite{finstermeier2013mitogenomic} and three other data sets (Anolis \cite{jackman1999phylogenetic}, RSV2 \cite{zlateva2004molecular,zlateva2005genetic} and HIV-1 \cite{shankarappa1999consistent}) in BEAST2 \cite{beast2data}.

The ESS and running time are summarised in Figure \ref{eff_comp} and Table \ref{eff_comp1}. To be more specific, we measure the efficiency by ESS per hour, which is calculated by the ESS of parameters in one simulation divided by the running time in hours. Then we compare the efficiency of two configurations by calculating the ratios of ESS per hour for simulations in the two configurations. If the ratio is larger than 1, then ESS per hour of Cons configuration is larger than that of Category configuration. As is shown in Figure \ref{eff_comp}, most ratios of the parameters are above the red line (larger than 1), which indicates that Cons configuration provides larger ESS per hour for most parameters. Although there are are several parameters sampled by Cons configuration having smaller ESS per hour in some data sets, it should be noticed that the ratio is calculated by choosing random simulations in the two configurations (See Appendix 3.3 for more details). Additionally, it is worth noting that the efficiency is improved more obviously in simulated data set having 1000 sites, compared with the data sets having 500 sites. This means the proposed operators sample rates and node times more efficiently if the genetic distances are more accurate. 

Table \ref{eff_comp1} lists the average running time of the data sets. It can be seen that Cons configuration finished simulations with less time in most cases. Moreover, Table \ref{eff_comp1} also shows the parameter that has the smallest ESS in Category configuration, and is compare with the corresponding ESS in Cons configuration. After calculating the ESS per hour, we conclude that Cons configuration improved the efficiency of the worst estimated parameter in Category configuration by a factor of 2.3 to 15.8. 

\subsection*{Correlation analysis of rates and branch lengths}
In this section, we conduct a pairwise comparison between rates and branch lengths in units of time. We used a data set of ratite mitochondrial genomes \cite{cooper2001complete}. This data set includes 7 species of ratites and an alignment of 10767 sites. After analysing the ratites data set in BEAST2 using the Constant Distance operator, we calculated the Pearson coefficient between the rates and the times across branches to investigate the posterior correlation of these parameters. 

The results are summarised in Figure \ref{correlation}. Figure \ref{correlation}(a) presents the topology of the maximum clade credibility tree. We utilised the programme TreeStat2 \cite{TreeStat2} to obtain the filtered trees that have the same topology as the maximum clade credibility tree from the sampled trees in MCMC chain. This means the trees that have different shared common ancestors of each taxon from the reference tree are filtered out. 

Afterwards, Figure \ref{correlation}(b) shows the pairwise comparison of the 12 branch rates and 12 branch lengths (in time) on these filtered trees. As can be seen from the diagonal, the rate on one branch is negatively correlated with the length of that branch, which indicates that an older divergence time will lead to a smaller rate. This is because the primary signal in the data is genetic distance, so that there will be a range of rates and divergence times that are consistent with the genetic distances, but the products of these quantities will vary less than the individual parameters. The consequence is that there will tend to be a negative relationship between rate ${r_i}$ and branch length ${l_i}$ i.e. ${r_i} = {d_i}/{l_i}$. At the same time, there will tend to be a positive relationship between rate ${r_i}$ and its parent's branch length ${l_{ip}}$, since a larger ${l_{ip}}$ leads to a smaller ${l_i}$. Moreover, for cherries that share the same branch length in the tree, they will tend to have the same correlation pattern. Take ANDI and DIGI as an example. ${r_1}$ and ${l_1}$ are negatively correlated, but ${r_1}$ and ${l_8}$ are positively correlated, which is also the correlation of ${r_2}$, ${l_2}$ and ${l_8}$. 

It is precisely this form of correlation structure in the posterior that our operator anticipates, and these correlations are the reason that our operator performs better than naive alternatives.

\subsection*{Sampling a fixed unrooted tree}
A limiting case for the relaxed molecular clock model (and one exploited in some of our validation tests) occurs for long sequences, when the branch lengths of the unrooted tree, in units of expected substitutions per site, become known without error. With full length genomes now available, although inferring trees from genomes involves complexities and assumptions such as a good partition scheme \cite{lanfear2012partitionfinder}, this limiting case might be approached in some data sets. As a simple test in this paper, this gives rise to an alternative approach to analysis, where divergence times, a root position and the branch rates are random variables, and the data are a set of branch lengths in units of substitution on a known unrooted tree topology. 

Previous work done by Reis and Yang \cite{reis2011approximate} also tried to approximate the likelihood of such an unrooted tree in Bayesian phylogenetic inference. Similar researches in \cite{thorne1998estimating,guindon2010bayesian} show that these methods can account for rate changes in a relaxed clock model, but the genetic distances are not fixed, for example St{\'e}phane Guindon used a Gibbs sampling algorithm \cite{guindon2010bayesian}. Outside of the Bayesian MCMC formalism, least-squares criteria \cite{to2015fast} and maximum likelihood \cite{sagulenko2018treetime,sanderson2003r8s}, can also be applied to estimate substitution rates and divergence times in unrooted trees. 

In this section, we investigated this approach on a fixed substitution tree reconstructed from whole mitochondrial genomes from a set of ratite species \cite{cooper2001complete}. Since no uncertainty is admitted in the genetic distances and the proposed operator don't change the genetic distances, the phylogenetic likelihood is no longer needed and the unrooted tree becomes the data, rather than a multiple sequence alignment.

First of all, we used the ratites data set to construct an unrooted tree with PhyML 3.0 \cite{phyml,guindon2010new}.  Figure \ref{withoutdata}(a) shows the unrooted tree with the genetic distances on the branches which are fixed in the subsequent relaxed clock analysis in BEAST2.

As an initial starting point, the root is assigned using the midpoint method. After that, according to the genetic distances among seven taxa and the position of the root, consistent divergence times are specified and assigned to each ancestral node, so that a valid rooted time tree is obtained. Once divergence times are determined, rates on the branches are also calculated so that the products match the unrooted substitution tree.

Then we used Constant Distance operator to sample a Markov chain initiated by this starting tree. The resulting posterior distribution is shown in Figure \ref{withoutdata}(b)-(d). As can be seen, despite that there is some uncertainty in the root position, the most probable tree in Figure \ref{withoutdata}(b) is consistent with previous analyses of this data (see Figure 2 in Ref. \cite{cooper2001complete}). For large data sets of long sequences, the proposed operators may prove useful to provide faster divergence time estimates based on the assumption of known unrooted topology and branch lengths in units of expected substitutions per site.

\section*{Discussion}
We have demonstrated that the presented operator is valid and able to improve the efficiency of phylogenetic MCMC for relaxed clock models. The overall performance of a Bayesian phylogenetic analysis will be affected by the proportion of MCMC steps that this operator is chosen to make the proposal. In the BEAST2 software, this can be changed by modifying the relative weights operators in the operator schedule. The ideal proportion is non-trivial to determine for an arbitrary data set. In this study, we assigned equal weights on operations to all internal nodes (including the root). How to assign weights to achieve better performance is not studied in this paper, and users may assign different weights in practice. Hence, an optimal method of assigning weights still needs further investigation. 

The key idea of the presented operator (to maintain the genetic distances) shows a novel direction for more efficient proposals in Bayesian phylogenetic MCMC. For example, the operations on the internal nodes, in the current study, involve one random internal node, one node time and three branch rates. If two or more nodes are selected, then more associated rates and node times can be sampled in one proposal, which may achieve even better efficiency. Another possible approach is to make small changes to the genetic distances as well. To minimise the number of changes to genetic distances, a two-dimensional random draw will be used to change four parameters (one divergence time and one rate changed directly, the other two rates derived so as to minimise changes to genetic distances). What's more, it should be pointed out that Small Pulley and Big Pulley can only be applied to reversible continuous-time Markov chain models where unrooted trees can be used in inference, because these operators require the underlying unrooted tree to be unchanged. Future work could elaborate a larger class of operators along these lines.

As data sets have increased in size the impetus to improve efficiency of Bayesian phylogenetic inference algorithms has steadily increased. Besides more effective proposal mechanisms within Metropolis-Hastings MCMC, completely novel approaches to Bayesian phylogenetics have also begun to get some attention. Variational methods are one alternative for approximating Bayesian posterior distributions \cite{beal2003variational}. These approaches make inference an optimisation problem and take advantage of tractable variational distributions that approximate the posterior distribution, thus decreasing the computational cost by avoiding high-dimensional integrals in MCMC sampling schemes. Recent work has investigated the potential for applying variational methods to phylogenetics \cite{zhang2018variational,dang2019stochastic}. Our improved MCMC methods provide a performance baseline for these new approaches.

\section*{Conclusions}
As data sets have increased in size, the need for computational efficiency of Bayesian phylogenetic analyses has also increased. In this paper, we have discussed a new tree proposal that substantially increases the efficiency of Bayesian phylogenetic inference under a popular class of relaxed molecular clock models.

We demonstrate the correctness of this algorithm with a series of tests including a well-calibrated simulation study. Based on both simulated and real data sets, the proposed operator is more efficient than current algorithms implemented in BEAST2. Performance improvements of greater than an order of magnitude increase in ESS/hour were measured on both real and simulated data. The proposed operator is available for use as a package of BEAST2.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{backmatter}
\section*{Abbreviations}
\begin{table}[h!]
\raggedright
\begin{tabular}{ll}
MCMC&Markov chain Monte Carlo\\
ESS&effective sample sizes\\ 
HPD&highest posterior density\\
Cateogry & Using the current operators in BEAST2 to sample discrete rate categories\\
Cons & Using the Constant Distance operator to sample continuous rates specified by an uncorrelated related clock model\\
\end{tabular}
\end{table}

\section*{Declarations}
\subsection*{Ethics approval and consent to participate}
Not applicable

\subsection*{Consent for publication}
Not applicable

\subsection*{Availability of data and material}
The source code of the proposed operator and the data sets analysed during the current study are available in the Github repository (https://github.com/Rong419/ConstantDistanceOperator.git).  

\subsection*{Competing interests}
The authors declare that they have no competing interests.

\subsection*{Funding}
The author(s) would like to acknowledge support from a Royal Society of New Zealand Marsden award (\#UOA1611; 16-UOA-277). This work is also partially supported by scholarship from China Scholarship Council (No.201706990021). 

\subsection*{Authors' contributions}
RZ developed the operator and was a major contributor in writing the manuscript. AJD supervised the implementation of the operator and the writing process of the manuscript. All authors read and approved the final manuscript.

\subsection*{Acknowledgements}
The author(s) wish to acknowledge the use of New Zealand eScience Infrastructure (NeSI) high performance computing facilities, consulting support and/or training services as part of this research. New Zealand's national facilities are provided by NeSI and funded jointly by NeSI's collaborator institutions and through the Ministry of Business, Innovation \& Employment's Research Infrastructure programme. URL https://www.nesi.org.nz.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{bmc_article}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figure Legends}
\begin{figure}[h!]
\includegraphics[width=12cm]{Fig01-flowchart.eps}\\
\caption{\csentence{The flow chart of the Constant Distance operator.}}
\label{flowchart}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig02-internalnode.eps}\\
\caption{\csentence{Illustration of the operation on an internal node.}
             The operator proposes ${t_X}'$, ${r_X}'$, ${r_L}'$ and ${r_R}'$, during which $d_L$, $d_R$, $d_X$ are kept constant.}
\label{internalnodes}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig03-rootstrategy.eps}\\
\caption{\csentence{Illustration of operations on root.}
             (a) An example of a 4-taxa unrooted tree and two possible rooted trees for the operator to sample, during which the unrooted tree can not be changed. Based on the original tree in (b), Simple Distance proposes a node time in ${g}'$ and two rates in $\bf{r}'$ and keeps $d_L$, $d_R$ constant in (c). Small Pulley proposes two rates in $\bf{r}'$ and $D = {d_L} + {d_R}$ remains constant in (d).}
\label{simpledistance}
\end{figure}


\begin{figure}[h!]
\includegraphics[width=12cm]{Fig04-exchangemethod.eps}\\
\caption{\csentence{Illustration of \textit{Exchange (\textbf{M},\textbf{N})} method.}
             This method is applied to tree $g$ and proposes $g'$ by swapping \textbf{M} and \textbf{N}, so that the three distances are adjusted to maintain the distances among \textbf{S}, \textbf{M} and \textbf{N}. That is, ${d_C}' = {d_C} + b$, ${d_N}' = d_C + d_N$ and ${d_M}' = d_{M} - {d_C}'$, where $b \sim U[ - v, + v]$.}
\label{exchangemethod}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig05-treeshape.eps}\\
\caption{\csentence{Two different tree shapes.}
             The symmetric tree is on the left and the asymmetric tree is on the right. The dashed triangles represent the potential subtrees rooted at the nodes.}
\label{treeshape}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig06-symmetric.eps}\\
\caption{\csentence{Illustration of operations on the symmetric tree in Figure \ref{treeshape}.}
             The proposed operator will propose one of the four possible trees, each with 0.25 probability.}
\label{symmetric}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig07-asymmetric.eps}\\
\caption{\csentence{Illustration of operations on the asymmetric tree in Figure \ref{treeshape}.}
             The proposed operator will propose one of the three possible trees. If ${t_o}' < t_{G1}$, \textcircled7 has 1 probability, otherwise \textcircled5 and \textcircled6 have 0.5 probability each.}
\label{asymmetric}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig08-modelvalidation.eps}\\
\caption{\csentence{The models and prior distributions to simulate the sequence data.}
             The sequence alignment (SA) is simulated through a phylogenetic continuous-time Markov Chain (PhyloCTMC) that consists of a substitution model (HKY) and an uncorrelated relaxed clock model (UCRelaxedClockModel). The random variables in HKY model construct the mutation rate matrix ($Q$), including base frequencies (${\bf{\pi}} = \{{\pi_A}, {\pi_C}, {\pi_G}, {\pi_T}\}$) and kappa ($\kappa$). The time trees ($\psi$) and branch rates ($r_i$ for each branch $i$ in $\psi$) construct the substitution tree (ST). The branch rates have a LogNormal prior with fixed mean 1 and certain standard deviation (denoted by ${s_1}$). And the time trees have a Yule model prior with birth rate ($\lambda$) having a LogNormal prior. The other prior distributions include a Dirichlet distributions on ${\bf{\pi}}$, a LogNormal distribution on $\kappa$, and a LogNormal distribution on ${s_1}$. For notations in LogNormal distributions, the uppercase letters represent the parameters in real space, and the lowercase letters represent the parameters in log space. In all the simulations, the number of taxa is fixed at 120 (n = 120).}
\label{modelvalidation}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig09-calibrated.eps}\\
\caption{\csentence{Well-calibrated simulation study with 120 taxa.} Each point is a separate simulated dataset.
             }
\label{LargeTree}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig10-efficiencycompare.eps}\\
\caption{\csentence{Comparison of ESS and running time.}
                   There are 6 data sets analysed, including 4 real data sets and 2 simulated data sets with different number of sites, as is shown in the legend. The red line represent the position where the ratio of ESS per hour is equal to 1. The horizontal axis represents the names of sampled parameters.}
\label{eff_comp}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig11-correlation.eps}\\
\caption{\csentence{Correlation analysis in the ratites tree.}
             $l$ represents the length of a branch, that is the time difference between a parent node and a child node, where ${l_1} = {l_2} =  {t_1} - 0$, ${l_3} = {l_4} = {t_2} - 0$, ${l_5} = {t_3} - 0$, ${l_6} = {t_4} - 0$, ${l_7} = T - 0$, ${l_8} = {t_5} - {t_1}$, ${l_9} = {t_3} - {t_2}$, ${l_{10}} = {t_4} - {t_3}$, ${l_{11}} = {t_5} - {t_4}$ and ${l_{12}} = T - {t_5}$. The rates and branch lengths are converted into log space and then Pearson's coefficients are computed, which range from -1 to 1. Blue indicates positive correlations and red indicates negative correlations. The darker the colour, the stronger the correlation. }
\label{correlation}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig12-fixedtree.eps}\\
\caption{\csentence{Illustration of sampling a fixed unrooted tree.}
            In subfigure (a), the unrooted tree is obtained from the ratites data set \cite{cooper2001complete} by a maximum likelihood method \cite{phyml} and the labeled numbers represent genetic distances. The three unique tree topologies in (b) (c) and (d) are obtained from the sampled trees by using program TreeTraceAnalysis \cite{TreeTraceAnalysis}. The branch rates and node times are summarised by using program TreeAnnotator \cite{TreeAnnotator}. The labeled numbers represent the posterior mean of rates on the corresponding branches. The colour of branches from green to red indicates the rates increasing from small to large, and the blue bars represent the 95\% HPD of the corresponding node times.}
\label{withoutdata}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig13-sampleprior.eps}\\
\caption{\csentence{The illustration of sampling from prior.}
             $g_1$ is set to be the original tree where an MCMC chain starts. When testing Big Pulley, the proposed operator samples the trees among $g_1$, $g_2$ and $g_3$.}
\label{sampleprior}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig14-testinternalnode.eps}\\
\caption{ \csentence{Sampled parameters in tests of internal nodes.}
             The horizontal axis represents the node time of D in Figure \ref{sampleprior}. The two scenarios sample two trees with different distances specified in Table \ref{ini_inter}.}
\label{res_int}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig15-testroot.eps}\\
\caption{\csentence{Sampled parameters in test of the root.}
             For the trees in Figure \ref{sampleprior}, Simple Distance samples the root time $t_E$ only, Small Pulley samples the distance $d_D$ only, and Big Pulley samples $t_E$, $t_D$, $d_D$. To make it simple, $t_E$ and $d_D$ are compared.}
\label{res_roo1}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig16-simulateddatatree.eps}\\
\caption{\csentence{The tree used to simulate sequence alignment.}
             The taxa are denoted by t1 to t20. The divergence times are drawn near the node.}
\label{simulateddatatree}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig17-AnolisTimeEss.eps}\\
\caption{\csentence{Running time and ESS using Anolis data.}}
\label{anolistimeess}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig18-RSV2TimeEss.eps}\\
\caption{\csentence{Running time and ESS using RSV2 data.}}
\label{rsv2timeess}
\end{figure}

\clearpage
\begin{figure}[h!]
\includegraphics[width=12cm]{Fig19-ShankarappaTimeEss.eps}\\
\caption{\csentence{Running time and ESS using HIV-1 data.}}
\label{shankarappatimeess}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig20-PrimatesTimeEss.eps}\\
\caption{\csentence{Running time and ESS using primates data.}}
\label{primatestimeess}
\end{figure}
\clearpage

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig21-SimulatedTimeEss.eps}\\
\caption{\csentence{Running time and ESS using simulated data.}}
\label{simulatedtimeess}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig22-AnolisProposal.eps}\\
\caption{\csentence{Efficiency comparison of proposals using Anolis data.}}
\label{anolisproposal}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig23-RSV2Proposal.eps}\\
\caption{\csentence{Efficiency comparison of proposals using RSV2 data.}}
\label{rsv2proposal}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=12cm]{Fig24-ShankarappaProposal.eps}\\
\caption{\csentence{Efficiency comparison of proposals using HIV-1 data.}}
\label{shankarappaproposal}
\end{figure}

\clearpage
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Use of \listoftables is discouraged.
%%
\section*{Tables}
\begin{table}[h!]
  \centering
\begin{tabular}{cc|cc}
\hline
Parameters&Coverage&Parameters&Coverage\\
\hline
Tree height&89&Ucldstdev&91\\
Tree length&91&${\pi_A}$&94\\
Kappa&97&${\pi_C}$&96\\
Birth rate&99&${\pi_G}$&95\\
Rate mean&100&${\pi_T}$&97\\
\hline
\end{tabular}
\caption{Percentage of real values lying in the 95\% HPD in Figure \ref{LargeTree} }\label{num_hpd}
\end{table}

\begin{table}[h!]
  \centering
\begin{tabular}{c|c|c|c|c}
\hline
Data&Configuration&Average running time&Parameter&ESS\\
\hline
\multirow{2}*{Anolis}&Category&0.9053&\multirow{2}*{rate.coeff}&486.32\\
&Cons&0.6231&&2237.97\\
\hline
\multirow{2}*{RSV2}&Category&5.1009&\multirow{2}*{rate.coeff}&383.98\\
&Cons&4.4077&&2379.33\\
\hline
\multirow{2}*{HIV-1}&Category&5.3436&\multirow{2}*{prior}&360.86\\
&Cons&5.2128&&795.21\\
\hline
\multirow{2}*{Primates}&Category&12.5615&\multirow{2}*{ucld.stdev}&57.47\\
&Cons&12.3820&&164.52\\
\hline
\multirow{2}*{Simualted 500 sites}&Category&0.4644&\multirow{2}*{rate.coeff}&793.27\\
&Cons&0.4805&&3047.09\\
\hline
\multirow{2}*{Simualted 1000 sites}&Category&1.9270&\multirow{2}*{rate.coeff}&1599.42\\
&Cons&0.4805&&6255.03\\
\hline
\end{tabular}
\caption{Summary of ESS and running time}\label{eff_comp1}
\end{table}

\begin{table}[h!]
  \centering
\begin{tabular}{c|cccc|c|c|cccc}
  \hline
&\multicolumn{4}{c|}{genetic distances (fixed)}&$t_D$&$t_E$&\multicolumn{4}{c}{initial rates}\\
&${d_j}$&${d_k}$&${d_x}$&${d_i}$&initial&(fixed)&${r_j}$&${r_k}$&${r_x}$&${r_i}$\\
\hline
Scenario 1&0.1&0.2&0.4&0.27&1&10&0.1&0.2&0.04&0.03\\
\hline
Scenario 2&0.4&0.8&2.4&1.6&0.4&0.8&1&2&3&4\\
  \hline
\end{tabular}
\caption{Initial settings for testing operations on internal nodes}\label{ini_inter}
\end{table}

\begin{table}[h!]
  \centering
\begin{tabular}{c|c|ccc|ccc|c}
  \hline
&\multirow{2}*{Chain Length}&\multicolumn{3}{c|}{Sample from MCMC}&\multicolumn{3}{c|}{Integral curve}&Plot\\
&~&Mean&Err&St.dev&Mean&Err&St.dev&\\
\hline
\multirow{2}*{Senario 1}&10000000&3.2727&8.3e-3&0.5467&\multirow{2}*{3.2669}&\multirow{2}*{1.3e-06}&\multirow{2}*{0.5553}&Figure \ref{res_int}(a)\\
~&20000000&3.271&6.1e-3&0.5616&~&~&~&Figure \ref{res_int}(b)\\
\hline
\multirow{2}*{Senario 2}&10000000&0.4677&3.9e-04&0.0265&\multirow{2}*{0.4667}&\multirow{2}*{3.5e-05}&\multirow{2}*{0.0262}&Figure \ref{res_int}(c)\\
~&20000000&0.4672&2.8e-04&0.0262&~&~&~&Figure \ref{res_int}(d)\\
  \hline
\end{tabular}
\caption{Results of sampling the internal node}\label{res_inter}
\end{table}

\begin{table}[h!]
  \centering
\begin{tabular}{c|cccc|c|c|cccc}
  \hline
\multirow{2}*{Strategy}&\multicolumn{4}{c|}{genetic distances}&\multirow{2}*{$t_D$}&\multirow{2}*{$t_E$}&\multicolumn{4}{c}{initial rates}\\
~&${d_j}$&${d_k}$&${d_x}$&${d_i}$&~&~&${r_j}$&${r_k}$&${r_x}$&${r_i}$\\
\hline
Simple Distance&0.1&0.2&0.4&0.27&1&10&0.1&0.2&0.04&0.03\\
Small Pulley&0.1&0.2&\multicolumn{2}{|c|}{0.67}&1&10&0.1&0.2&0.04&0.03\\
Big Pulley&0.5&0.5&\multicolumn{2}{|c|}{0.5}&5&10&0.1&0.1&0.03&0.04\\
  \hline
\end{tabular}
\caption{Initial settings for operations on the root}\label{ini_sim}
\end{table}

\begin{table}[h!]
\centering
\begin{tabular}{c|c|cc|cc|cc}
  \hline
\multirow{2}*{Strategy}&\multirow{2}*{Variable}&\multicolumn{2}{c|}{Sample from MCMC}&\multicolumn{2}{c|}{Integral curve}&\multirow{2}*{Plot}\\
&~&Mean&St.dev&Mean&St.dev&\\
\hline
Simple Distance&$t_E$&7.8081&1.2884&7.8187&1.2992&Figure \ref{res_roo1}(a)\\
\hline
Small Pulley&${d_i}$&0.3480&0.0492&0.3476&0.0494&Figure \ref{res_roo1}(b)\\
\hline
\multirow{2}*{Big Pulley}&${d_i}$&0.1016&0.0766&0.0960&0.0760&Figure \ref{res_roo1}(c)\\
~&$t_E$&3.3017&0.6908&3.3095&0.6912&Figure \ref{res_roo1}(d)\\
\hline
\end{tabular}
\caption{Results of sampling the root}\label{res_sma}
\end{table}

%\begin{adjustbox}{angle=90}
\begin{sidewaystable}
\caption{Operator weights in MCMC chains}\label{weights}
\centering
\begin{tabular}{c|c|cc|cc|cc|cc|cc}
  \hline
\multirow{2}*{Operator class}&\multirow{2}*{Name}&\multicolumn{2}{c|}{Simulated data}&\multicolumn{2}{c|}{Anolis}&\multicolumn{2}{c|}{RSV2}&\multicolumn{2}{c|}{HIV-1}&\multicolumn{2}{c}{Primates}\\
&&Cons&Category&Cons&Category&Cons&Category&Cons&Category&Cons&Category\\
  \hline
\multirow{8}*{rates times}&ConstantDistance Operator&0.1919&-&0.2248&-&0.2228&-&0.2228&-&0.1850&-\\
&Rate Normal Operators$^1$&0.1919&0.2879&0.1349&0.2698&0.1337&0.2674&0.1337&0.2674&0.1850&0.2775\\
&UcldStdev Scale Operator$^2$&0.0288&0.0288&0.0270&0.0270&0.0267&0.0267&0.0267&0.0267&0.0278&0.0278\\
&UcldMean Scale Operator&-&-&-&-&0.0357&0.0089&0.0089&0.0089&-&-\\
&UcldMean Tree UpperDown Operator&-&-&-&-&0.0446&0.0267&0.0267&0.0267&-&-\\
&InternalNodeTime Scale Operator&0.0480&0.0960&0.0270&0.0270&0.0267&0.0267&0.0267&0.0267&0.0463&0.0463\\
&RootAge Scale Operator&0.0480&0.0480&0.0270&0.0270&0.0267&0.0267&0.0267&0.0267&0.0463&0.0463\\
&AllNodeTimes Uniform Operator&0.0480&0.0960&0.1799&0.2698&0.1337&0.2674&0.1783&0.2674&0.1850&0.2775\\
  \hline
\multirow{7}*{Tree}&SubtreeSlide Operator&0.1440&0.1440&0.0989&0.1349&0.1070&0.1337&0.1159&0.1337&0.1388&0.1388\\
&NarrowExchange Operator&0.1440&0.1440&0.0989&0.1349&0.1070&0.1337&0.1159&0.1337&0.0463&0.0463\\
&WideExchange Operator&0.0480&0.0480&0.0270&0.0270&0.0267&0.0267&0.0267&0.0267&0.0463&0.0463\\
&WilsonBalding Operator&0.0480&0.0480&0.0270&0.0270&0.0267&0.0267&0.0267&0.0267&0.0463&0.0463\\
&BirthRate Scale Operator&0.0480&0.0480&0.0629&0.0270&-&-&-&-&0.0278&0.0278\\
&DeathRate Scale Operator&-&-&0.0629&0.0270&-&-&-&-&-&-\\
&PopulationSize Scale Operator&-&-&-&-&0.0802&0.0267&0.0624&0.0267&-&-\\
  \hline
\multirow{2}*{Substituion model}&Kappa Scale Operator&0.0096&0.0096&0.0009&0.0009&0.0009&0.0009&0.0009&0.0009&0.0185&0.0185\\
&Frequencies DeltaExchange Operator&0.0019&0.0019&0.0009&0.0009&0.0009&0.0009&0.0009&0.0009&0.0009&0.0009\\
  \hline
\multicolumn{12}{l}{Note}\\
\multicolumn{12}{l}{1: Random walk operator and Swap operator in Cons configuration, Random walk operator, Scale operator and Swap operator in Category configuration.}\\
\multicolumn{12}{l}{2: The operator introduced in Appendix section 4 is used in Cons configuration, a Scale operator is used in Category configuration
.}\\
\multicolumn{12}{l}{-: The parameter is not sampled and no operator is assigned.}\\
\end{tabular}
\end{sidewaystable}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\clearpage
\newpage
\section*{Appendix}

\subsection*{1. The Green ratio}

When developing an operator for MCMC, the proposal function must be reversible. In other words, the probability that the operator propose a new state from the current state is required to be equal to the probability that the proposed state goes back to current state. To be specific, let ${\pi (x)}$ be the target probability distribution and $p(x, x')$ be the transition kernel in the continuous Markov chain. The reversibility condition requires that ${\pi (x)}{p(x, x')} = {\pi (x')}{p(x', x)}$. And an operator provides a proposal $q(x, x')$ with some probability $\alpha(x, x')$ that the proposal is accepted. Thus, the reversibility condition is rewritten as ${\pi (x)}{q(x, x')}{\alpha(x, x')} = {\pi (x')}{q(x', x)}{\alpha(x', x)}$.

Considering the subspace $\varphi_1$ on $x$ and subspace $\varphi_2$ on $x'$, it is assumed that there is a symmetric measure on the combined parametric space $\varphi = {\varphi_1} \times {\varphi_2}$, so that ${\pi (x)}{q(x, x')}$ has a density with respect to a single measure on $\varphi$. Then, Green suggested that the reversibility condition should be satisfied by detailed balance \cite{green1995reversible}, as represented by equation (\ref{balance}). And according to Peskun' proof, it is optimal to take equation (\ref{Ratio}) as the acceptance probability to retain the detailed balance \cite{peskun1973optimum}.

\begin{equation}\label{balance}
\int_A {\pi (x) d_x} {\int_B  {q(x, x')}{\alpha(x, x')} d_x}  = \int_B {\pi (x') d_{x'}}{\int_A {q(x', x)}{\alpha(x', x)} d_{x'}} \text{,}
\end{equation}
where $A \in {\varphi_1}$ and $B \in  {\varphi_2}$ are two Borel sets. ${q(x, x')}$ denotes the probability that the operator proposes a new state $x'$ given the current state $x$.

\begin{equation}\label{Ratio}
{\alpha_H}(x, x') = \min \left\{ {1,\frac{{\pi (x'){p}(x',x)}}{{\pi (x){p}(x,x')}}} \right\} \text{,}
\end{equation}
where ${p(x',dx)}/{p(x,dx')}$ is known as the Hastings ratio.

However, for operators that do not have a symmetric measure, it is necessary to include the Jacobian matrix ${\mathbf{J}}$  in order to deal with the dimension matching problem, as is discussed in Green's paper \cite{green1995reversible}. In this case, equation (\ref{Ratio}) is extended, as is shown in equation (\ref{Green}).
\begin{equation}\label{Green}
{\alpha_G}(x, x') = \min \left\{ {1,\frac{{\pi (x'){p}(x',x)}}{{\pi (x){p}(x,x')}}}\left|{\mathbf{J}}\right| \right\} \text{,}
\end{equation}
where ${\mathbf{J}} = {\nabla h(x, x')}$ represents a vector differential matrix of deterministic function $h$. $\alpha = \frac{{p}(x',x)}{{p}(x,x')}\left|{\mathbf{J}}\right|$ is defined as the Green ratio, and ${\mathbf{J}}$ ensures that the proposal have a symmetric measure on each subspace in state $x$ and $x'$.

\subsection*{1.1 Calculating the Green ratio for operations on internal nodes}

The Constant Distance Operator firstly proposes a new time for the randomly selected internal node (equation (\ref{in_HR1.1})), and then proposes three rates by the original distances and new node times(equation (\ref{in_HR1.2})$\sim$equation (\ref{in_HR1.3})).
\begin{subequations}\label{in_HR1}
\begin{equation}\label{in_HR1.1}
{f_1}:{{\text{t}}_X}{\text{'  =  }}{{\text{t}}_X}{\text{  +  a}}
\end{equation}
\begin{equation}\label{in_HR1.2}
{f_2}:{r_X}' = \frac{{{r_X} \times ({t_P} - {t_X})}}{{{t_P} - {t_X}'}}
\end{equation}
\begin{equation}
{f_3}:{r_L}' = \frac{{{r_L} \times ({t_X} - {t_L})}}{{{t_X}' - {t_L}}}
\end{equation}
\begin{equation}\label{in_HR1.3}
{f_4}:{r_R}' = \frac{{{r_R} \times ({t_X} - {t_2})}}{{{t_X}' - {t_R}}}
\end{equation}
\end{subequations}

Substituting equation (\ref{in_HR1}) in the Jacobian matrix ${{\mathbf{J}}_1}$ (equation (\ref{JacobianMatrix})), we can get equation (\ref{in_HR2}), so that the determinant of ${{\mathbf{J}}_1}$ can be obtained by equation (\ref{in_HR3}).
\begin{equation}\label{in_HR2}
{{\mathbf{J}}_1} = \left[ {\begin{array}{*{20}{c}}
  1&0&0&0 \\
  {\frac{{ - {r_X}}}{{{t_P} - {t_X}'}}}&{\frac{{{t_P} - {t_X}}}{{{t_P} - {t_X}'}}}&0&0 \\
  {\frac{{{r_L}}}{{{t_X}' - {t_L}}}}&0&{\frac{{{t_X} - {t_L}}}{{{t_X}' - {t_L}}}}&0 \\
  {\frac{{{r_R}}}{{{t_X}' - {t_R}}}}&0&0&{\frac{{{t_X} - {t_R}}}{{{t_X}' - {t_R}}}}
\end{array}} \right]
\end{equation}
\begin{equation}\label{in_HR3}
\begin{aligned}
\left| {{{\mathbf{J}}_1}} \right| &= 1 \times \left| {\begin{array}{*{20}{c}}
  {\frac{{{t_P} - {t_X}}}{{{t_P} - {t_X}'}}}&0&0 \\
  0&{\frac{{{t_X} - {t_L}}}{{{t_X}' - {t_L}}}}&0 \\
  0&0&{\frac{{{t_X} - {t_R}}}{{{t_X}' - {t_R}}}}
\end{array}} \right| \\&= \frac{{{t_P} - {t_X}}}{{{t_P} - {t_X}'}} \times \left| {\begin{array}{*{20}{c}}
  {\frac{{{t_X} - {t_L}}}{{{t_X}' - {t_L}}}}&0 \\
  0&{\frac{{{t_X} - {t_R}}}{{{t_X}' - {t_R}}}}
\end{array}} \right| \\&= \frac{{{t_P} - {t_X}}}{{{t_P} - {t_X}'}} \times \frac{{{t_X} - {t_L}}}{{{t_X}' - {t_L}}} \times \frac{{{t_X} - {t_R}}}{{{t_X}' - {t_R}}}
\end{aligned}
\end{equation}

\subsection*{1.2 Calculating the Green ratio for Simple Distance}

Simple Distance proposes two rates by using equation (\ref{SD_HR1.2}) and equation (\ref{SD_HR1.3}), according the new root time in equation (\ref{SD_HR1.1}). So the Jacobian matrix can be obtained as is shown in equation (\ref{SD_HR2}).
\begin{subequations}\label{SD_HR1}
\begin{equation}\label{SD_HR1.1}
 {t_X}' = {t_X} + a
\end{equation}
\begin{equation} \label{SD_HR1.2}
{r_L}' = \frac{{{r_L} \times ({t_X} - {t_L})}}{{{t_X}' - {t_L}}}
\end{equation}
\begin{equation}\label{SD_HR1.3}
  {r_R}' = \frac{{{r_R} \times ({t_X} - {t_R})}}{{{t_X}' - {t_R}}}
\end{equation}
\end{subequations}
\begin{equation}\label{SD_HR2}
{{\mathbf{J}}_2} = \left[ {\begin{array}{*{20}{c}}
  {\frac{{\partial {t_X}'}}{{\partial {t_X}}}}&{\frac{{\partial {t_X}'}}{{\partial {r_X}}}}&{\frac{{\partial {t_X}'}}{{\partial {r_R}}}} \\
  {\frac{{\partial {r_L}'}}{{\partial {t_X}}}}&{\frac{{\partial {r_L}'}}{{\partial {r_X}}}}&{\frac{{\partial {r_L}'}}{{\partial {r_R}}}} \\
  {\frac{{\partial {r_X}'}}{{\partial {t_X}}}}&{\frac{{\partial {r_X}'}}{{\partial {r_X}}}}&{\frac{{\partial {r_X}'}}{{\partial {r_R}}}}
\end{array}} \right] = \left[ {\begin{array}{*{20}{c}}
  1&0&0 \\
  {\frac{{{r_L}}}{{{t_X}' - {t_L}}}}&{\frac{{{t_X} - {t_L}}}{{{t_X}' - {t_L}}}}&0 \\
  {\frac{{{r_x}}}{{{t_X}' - {t_R}}}}&0&{\frac{{{t_X} - {t_R}}}{{{t_X}' - {t_R}}}}
\end{array}} \right]
\end{equation}

So the determinant of ${{\mathbf{J}}_2}$ is calculated by equation (\ref{SD_HR3})
\begin{equation}\label{SD_HR3}
\left| {{{\mathbf{J}}_2}} \right| = \frac{{{t_X} - {t_L}}}{{{t_X}' - {t_L}}} \times \frac{{{t_X} - {t_R}}}{{{t_X}' - {t_R}}}
\end{equation}
\subsection*{Calculating the Green ratio for Small Pulley}

Small Pulley proposes a new genetic distance of a branch on one side of the root by adding a random number $b$, which is equal to adding a random number $b$ to the original product of rate and time on that branch. As a result, a new rate is proposed by equation (\ref{SP_HR1.1}). Similarly, a new rate on another branch is proposed by equation (\ref{SP_HR1.2}), because the total distance of the two branches linked to the root should remain constant.
\begin{subequations}\label{SP_HR1}
\begin{equation}\label{SP_HR1.1}
{r_L}' = \frac{{{r_L} \times ({t_X} - {t_L}) + b}}{{{t_X} - {t_L}}}
\end{equation}
\begin{equation}\label{SP_HR1.2}
{r_R}' = \frac{{[{r_R} \times ({t_X} - {t_R}) + {r_L} \times ({t_X} - {t_L})] - [{r_L} \times ({t_X} - {t_L}) + b]}}{{{t_X} - {t_R}}} = \frac{{{r_R} \times ({t_X} - {t_R}) - b}}{{{t_X} - {t_R}}}
\end{equation}
\end{subequations}

Then, as is illustrated in equation (\ref{SP_HR2}), the Jacobian matrix ${{\mathbf{J}}_3}$ is simply obtained, which makes the determinant $\left| {{{\mathbf{J}}_3}} \right| = 1$.
\begin{equation}\label{SP_HR2}
{{\mathbf{J}}_3} = \left[ {\begin{array}{*{20}{c}}
  {\frac{{\partial {r_L}'}}{{\partial {r_L}}}}&{\frac{{\partial {r_L}'}}{{\partial {r_X}}}} \\
  {\frac{{\partial {r_R}'}}{{\partial {r_L}}}}&{\frac{{\partial {r_X}'}}{{\partial {r_X}}}}
\end{array}} \right] = \left[ {\begin{array}{*{20}{c}}
  1&0 \\
  0&1
\end{array}} \right]
\end{equation}
\subsection*{1.3 Calculating the Green ratio for Big Pulley}

Two new node times are proposed in Big Pulley. One is the root time (equation (\ref{BP_HR1.1})), the other is the node time of the child node of the root. It can be either children of the root, i.e. \textbf{son} and \textbf{dau}. So ${t_C}'$ is used to denote the node time proposed, as is seen in equation (\ref{BP_HR1.2}). In addition, the distances are adjusted by the method \textit{Exchange (\textbf{M}, \textbf{N})}, dependent on which nodes are chosen. As a result, the four rates are proposed, as is shown in equation (\ref{BP_HR1.3})$\sim$equation (\ref{BP_HR1.6})
\begin{subequations}\label{BP_HR1}
\begin{equation}\label{BP_HR1.1}
{t_X}' = {t_X} + a
\end{equation}
\begin{equation}\label{BP_HR1.2}
{t_C}' = {t_C} + {a_{1,2,3}}
\end{equation}
\begin{equation}\label{BP_HR1.3}
{r_C}' = \frac{{{r_C} \times (t{}_X - {t_C}) + b}}{{t{}_X' - {t_C}'}}
\end{equation}
\begin{equation}\label{BP_HR1.4}
{r_S}' = \frac{{{r_2} \times (t{}_C - {t_S})}}{{t{}_C' - {t_S}}}
\end{equation}
\begin{equation}\label{BP_HR1.5}
{r_{M}}' = \frac{{{r_{M}} \times ({t_C} - {t_{M}}) - [{r_C} \times ({t_X} - {t_C}) + b]}}{{{t_X}' - {t_{M}}}}
\end{equation}
\begin{equation}\label{BP_HR1.6}
{r_{N}}' = \frac{{{r_C} \times ({t_X} - {t_C}) + {r_{N}} \times ({t_X} - {t_{N}})}}{{{t_C}' - {t_{N}}}}
\end{equation}
\end{subequations}
where ${a_{1,2,3}}$ is the random number to propose a new node time for the child node of the root. Depending on which child node is selected, the notation is different, i.e. ${a_1}$, ${a_2}$, ${a_3}$. Here, to make it a general case, ${a_x}$ is used.

Therefore, the Jacobian matrix ${{\mathbf{J}}_4}$ for the six parameters in equation (\ref{BP_HR1}) is obtained by equation (\ref{BP_HR2}). And the determinant of ${{\mathbf{J}}_4}$ is calculated shown in equation (\ref{BP_HR3}).
\begin{equation}\label{BP_HR2}
\begin{aligned}
{{\mathbf{J}}_4} &= \left[ {\begin{array}{*{20}{c}}
  {\frac{{\partial {t_X}'}}{{\partial {t_X}}}}&{\frac{{\partial {t_X}'}}{{\partial {t_C}}}}&{\frac{{\partial {t_X}'}}{{\partial {r_C}}}}&{\frac{{\partial {t_X}'}}{{\partial {r_S}}}}&{\frac{{\partial {t_X}'}}{{\partial {r_{M}}}}}&{\frac{{\partial {t_X}'}}{{\partial {r_{N2}}}}} \\
  {\frac{{\partial {t_C}'}}{{\partial {t_X}}}}&{\frac{{\partial {t_C}'}}{{\partial {t_C}}}}&{\frac{{\partial {t_C}'}}{{\partial {r_C}}}}&{\frac{{\partial {t_C}'}}{{\partial {r_S}}}}&{\frac{{\partial {t_C}'}}{{\partial {r_{M}}}}}&{\frac{{\partial {t_C}'}}{{\partial {r_{N2}}}}} \\
  {\frac{{\partial {r_C}'}}{{\partial {t_X}}}}&{\frac{{\partial {r_C}'}}{{\partial {t_C}}}}&{\frac{{\partial {r_C}'}}{{\partial {r_C}}}}&{\frac{{\partial {r_C}'}}{{\partial {r_S}}}}&{\frac{{\partial {r_C}'}}{{\partial {r_{M}}}}}&{\frac{{\partial {r_C}'}}{{\partial {r_{N2}}}}} \\
  {\frac{{\partial {r_S}'}}{{\partial {t_X}}}}&{\frac{{\partial {r_S}'}}{{\partial {t_C}}}}&{\frac{{\partial {r_S}'}}{{\partial {r_C}}}}&{\frac{{\partial {r_S}'}}{{\partial {r_S}}}}&{\frac{{\partial {r_S}'}}{{\partial {r_{M}}}}}&{\frac{{\partial {r_S}'}}{{\partial {r_{N2}}}}} \\
  {\frac{{\partial {r_{M}}'}}{{\partial {t_X}}}}&{\frac{{\partial {r_{M}}'}}{{\partial {t_C}}}}&{\frac{{\partial {r_{M}}'}}{{\partial {r_C}}}}&{\frac{{\partial {r_{M}}'}}{{\partial {r_S}}}}&{\frac{{\partial {r_{M}}'}}{{\partial {r_{M}}}}}&{\frac{{\partial {r_{M}}'}}{{\partial {r_{N}}}}} \\
  {\frac{{\partial {t_{N}}'}}{{\partial {t_X}}}}&{\frac{{\partial {t_{N}}'}}{{\partial {t_C}}}}&{\frac{{\partial {t_{N}}'}}{{\partial {r_C}}}}&{\frac{{\partial {t_{N}}'}}{{\partial {r_S}}}}&{\frac{{\partial {t_{N}}'}}{{\partial {r_{N}}}}}&{\frac{{\partial {t_{N}}'}}{{\partial {r_{N}}}}}
\end{array}} \right] \\&= \left[ {\begin{array}{*{20}{c}}
  1&0&0&0&0&0 \\
  0&1&0&0&0&0 \\
  {\frac{{{r_C}}}{{{t_X}' - {t_C}'}}}&{\frac{{ - {r_C}}}{{{t_X}' - {t_C}'}}}&{\frac{{{t_X}' - {t_C}}}{{{t_X}' - {t_C}'}}}&0&0&0 \\
  0&{\frac{{{r_S}}}{{t' - {t_S}}}}&0&{\frac{{{t_C} - {t_S}}}{{{t_C}' - {t_S}}}}&0&0 \\
  {\frac{{ - {r_C}}}{{{t_X}' - {t_{M}}}}}&{\frac{{{r_{N1}} + {r_C}}}{{{t_X}' - {t_{M}}}}}&{\frac{{ - ({t_X} - {t_C})}}{{{t_X}' - {t_{M}}}}}&0&{\frac{{{t_C} - {t_{M}}}}{{{t_X}' - {t_{M}}}}}&0 \\
  {\frac{{{r_C} + {r_S}}}{{{t_C}' - {t_{N}}}}}&{\frac{{ - ({r_C} + {r_S})}}{{{t_C}' - {t_{N}}}}}&{\frac{{{t_X} - {t_C}}}{{{t_C}' - {t_{N}}}}}&0&0&{\frac{{{t_X} - {t_{N}}}}{{{t_C}' - {t_{N}}}}}
\end{array}} \right]
\end{aligned}
\end{equation}
\begin{equation}\label{BP_HR3}
\left| {{{\mathbf{J}}_4}} \right| = \frac{{{t_X}' - {t_C}}}{{{t_X}' - {t_C}'}} \times \frac{{{t_C} - {t_S}}}{{{t_C}' - {t_S}}} \times \frac{{{t_C} - {t_{M}}}}{{{t_X}' - {t_{M}}}} \times \frac{{{t_X} - {t_{N}}}}{{{t_C}' - {t_{N}}}}
\end{equation}

Last but not least, due to the change of tree topology in \textit{Exchange (\textbf{M}, \textbf{N})}, the probability of the proposed tree going back to the original tree $p(g|g')$, as well as the probability of making the proposal $p(g'|g)$, should be considered. As the ratio of $p(g|g')/p(g'|g)$ is defined as $\mu$, the calculation of $\mu$ is detailed in the following algorithm.
\begin{algorithm}
\caption{Calculation of $\mu$ for Big pulley}
\label{alg1}
\begin{algorithmic}
\STATE Original tree is symmetric:
\IF{the node that has been exchanged with \textbf{L} or \textbf{R} has child nodes}
\STATE $\alpha  = \beta  = 0.25$
\ELSIF{${t_R} > {t_L}$}
\STATE $\alpha  = 1,\beta  = 0.5$
\ELSIF{${t_R} < {t_L}$}
\STATE $\alpha  = 0.5,\beta  = 1$
\ELSIF{${t_R} = {t_L}$}
\STATE $\alpha  = \beta  = 1$
\ENDIF
\IF{Proposed tree belongs to \textcircled1 or \textcircled2}
\STATE Return $\mu = \frac{\alpha }{{0.25}}$
\ENDIF
\IF{Proposed tree belongs to \textcircled3 or \textcircled4}
\STATE Return $\mu = \frac{\beta }{{0.25}}$
\ENDIF
\STATE
\STATE Original tree is asymmetric:
\IF{the node that has been exchanged with \textbf{O} has child nodes}
\STATE $\gamma  = 0.25$
\ELSE
\STATE $\gamma  = 0.5$
\ENDIF

\IF{Proposed tree belongs to \textcircled5 or \textcircled6}
\STATE Return $\mu = \frac{\gamma }{{0.5}}$
\ENDIF
\IF{Proposed tree belongs to \textcircled7}
\STATE Return $\mu = \frac{{0.25}}{1}$
\ENDIF

\end{algorithmic}
\end{algorithm}

\subsection*{2. Sampling from the prior}

In this section, we aim to validate the correctness of the proposed operators. To be more specific, we firstly run the simulations by sampling from prior distributions in BEAST2. Since the prior distributions are deterministic, we can analytically calculate the theoretical joint-distributions of sampled parameters in MCMC chains. By comparing the sampled distributions with the analytical results, we demonstrate whether the proposed operators are able to sample parameters correctly.

In Figure \ref{sampleprior}, a tree with three taxa $A$, $B$ and $C$ (plus one internal node $D$, and root $E$) is used as a small example in the experiments in this section. In the figure, $g_1$ is set as the initial tree. Firstly, a LogNormal distribution is used as the rate prior in the uncorrelated relaxed clock model, given by equation (\ref{rateprior}).
\begin{equation}\label{rateprior}
r = \{{r_A}\quad{r_B}\quad{r_C}\quad{r_D}\} \sim LogNormal(m = -3, s = 0.25)
\end{equation}

In addition, a Coalescent model \cite{pybus2002genie} with constant population size ($N=0.3$) is used to describe the tree prior. Hence, for the tree in Figure \ref{sampleprior}, the probability of node times is calculated by equation (\ref{treeprior}).
\begin{equation}\label{treeprior}
p(t=\{{t_E},{t_D}\}) = (\frac{1}{N} \times {e^{ - \frac{1}{N}({t_E} - {t_D})}}) \times (\frac{1}{N} \times {e^{ - \frac{3}{N}{t_D}}})
\end{equation}

After the priors are specified, the distribution to sample can be exactly known, since the samples are drawn from the prior distributions. In other words, as the rates are functions of its genetic distance and times, the joint distribution to sample can be represented by equation (\ref{integration}).
\begin{equation}\label{integration}
\begin{aligned}
p(r,t) &= p({t_E},{t_D}) \times p({r_D}) \times p({r_A}) \times p({r_B}) \times p({r_C}) \\&= p({t_E},{t_D}) \times p(\frac{{{d_D}}}{{t_E} - {t_D}}) \times p(\frac{{{d_A}}}{{t_D} - {t_A}}) \times p(\frac{{{d_B}}}{{t_D} - {t_B}}) \times p(\frac{{{d_C}}}{{t_E} - {t_C}})\text{,}
\end{aligned}
\end{equation}
where $p({.})$ is the probability of certain rate values in the LogNormal distribution. Therefore, the whole probability can be obtained by conducting numerical integration on equation (\ref{integration}), which shows the probability distribution over all the possible values of parameters.

\subsubsection*{2.1 Test the operator on internal nodes}

The genetic distances, node times and rates for $g_1$ in Figure \ref{sampleprior} are given in Table \ref{ini_inter}. To test roundly, two scenarios are designed. In each scenario, the genetic distances are fixed, the node time $t_D$ starts from the initial value and will be changed by the proposed operator during the sampling process. Essentially, the proposed operator makes node $D$ move between node $A$ and $E$. Besides, to make sure that the result is robust, two different MCMC chain lengths are performed in each scenario, i.e. 10 million and 20 million.

The mean, mean error and the standard deviation of the MCMC samples are summarised in Table \ref{res_inter}. Besides, according to equation (\ref{integration}), the actual joint distribution is obtained by using equation (\ref{integration1}), and is used to evaluate the results, which is also included in Table \ref{res_inter}. Moreover, the histograms of MCMC samples that indicate the sampled distributions, as well as the curves of the numerical integration of equation (\ref{integration1}), are shown in Figure \ref{res_int}. From Table \ref{res_inter} and Figure \ref{res_int}, it can be seen that the red curves well fit the black histograms, and the mean values and standard deviations are consistent, which makes it safe to conclude that the proposed operator samples the internal node correctly.
\begin{equation}
\label{integration1}
p(r,t) = \int_{{t_D} = 0}^{{t_E}} {p({t_E},{t_D}) \times p(\frac{{{d_A}}}{{{t_D}}}) \times p(\frac{{{d_B}}}{{{t_D}}}) \times p(\frac{{{d_D}}}{{{t_E} - {t_D}}}) \times p(\frac{{{d_C}}}{{{t_E}}}){d_{t_D}}}
\end{equation}

\subsubsection*{2.2 Test the operator on root}

Still starting from $g_1$ in Figure \ref{sampleprior}, the initial settings for testing the root are given in Table \ref{ini_sim}. And the three strategies are tested separately in the following parts.

\paragraph*{2.2.1 Using Simple Distance}

The root time $t_E$ is sampled by Simple Distance, which ranges from 1 to positive infinity theoretically. Namely, all the genetic distances and the node time $t_D$  are fixed. Similar to equation (\ref{integration1}), the joint distribution of $t_E$ and rates to sample can be obtained by equation (\ref{integration2}).
\begin{equation}
\label{integration2}
p(r,t) = \int_{{t_E} = 1}^{ + \infty } {p({t_E},{t_D}) \times p(\frac{{{d_A}}}{{{t_D}}}) \times p(\frac{{{d_B}}}{{{t_D}}}) \times p(\frac{{{d_D}}}{{{t_E} - {t_D}}}) \times p(\frac{{{d_C}}}{{{t_E}}}){d_{t_E}}}
\end{equation}

The results are given in Table \ref{res_sma} and Figure \ref{res_roo1}(a). As can be seen, the mean and the standard deviation of MCMC samples and numerical integration are close to each other, which confirms that the two distribution are the same. Thus, Simple Distance is proved to be correct.

\paragraph*{2.2.2 Using Small Pulley}

Although both ${d_x}$ and ${d_i}$ are changed during the sampling process when using Small Pulley, the sum of ${d_D}$ and ${d_C}$ are kept 0.67 in this test, as the initial setting shown in Table \ref{ini_sim}. To make it simple, only ${d_D}$ is compared.

Then, based on equation (\ref{integration}), the exact distribution of ${d_i}$ can be obtained by equation (\ref{integration3}), which is compared with the sampled distribution in Table \ref{res_sma} and Figure \ref{res_roo1}(b). Even though there exist some errors, the sampled parameters can be considered to follow the same distribution. So the Small Pulley is also able to provide correct samples.
\begin{equation}\label{integration3}
p(r,t) = \int_{{d_D} = 1}^{0.67} {p({t_E},{t_D}) \times p(\frac{{{d_A}}}{{{t_D}}}) \times p(\frac{{{d_B}}}{{{t_D}}}) \times p(\frac{{{d_D}}}{{{t_E} - {t_D}}}) \times p(\frac{{0.67 - {d_D}}}{{{t_E}}}){d_{d_D}}}
\end{equation}

\paragraph*{2.2.3 Using Big Pulley}

For $g_1$ in Figure \ref{sampleprior}, a new tree, together with the root time $t_E$ and node time of its older child $t_D$, as well as a genetic distance $d_i$, is proposed by Big Pulley. In this case, the initial tree $g_1$ will either go to $g_2$ or $g_3$, as is shown in Figure \ref{sampleprior}.  So the samples are repeatedly drawn from the 3 trees. Besides, according to the initial settings in Table \ref{ini_sim}, the genetic distances remain unchanged during the process, i.e. $d_{AB} = 1$, $d_{AC}  = 1$ and $d_{BC} = 1$ hold. Hence, the distribution we are about to achieve can be calculated by equation (\ref{integration5}).
\begin{equation}\label{integration5}
\begin{aligned}
p(r,t) &= \int_{{t_E} = 0}^{ + \infty } {\int_{{t_D} = 0}^{{t_E}} {\int_{{d_D} = 0}^{0.5} {p({t_E},{t_D})} } \times p(\frac{{0.5}}{{{t_D}}})}  \\&\times p(\frac{{0.5}}{{{t_D}}}) \times p(\frac{{{d_D}}}{{{t_E} - {t_D}}}) \times p(\frac{{0.5 - {d_D}}}{{{t_E}}}){d_{d_D}}{d_{t_D}}{d_{t_E}}
\end{aligned}
\end{equation}

The statistical measurements, i.e. mean and standard deviation, are compared in Table \ref{res_sma}. The histograms of samples and numerical curves of ${d_D}$ and ${t_E}$ are pictured in Figure \ref{res_roo1}(c) and Figure \ref{res_roo1}(d). It is shown that the two distributions are consistent within the acceptable error range. Therefore, Big Pulley can also give the right combinations of rates and node times, under the condition that the genetic distances among taxa are constant.

\subsection*{3. Performance analysis of operators}\label{performanceanalysis}
This section provides the details of the results presented in \textit{Performance comparison} section.

\paragraph*{3.1 Operator weights}

The weights on operators for the simulations when comparing efficiency are listed in Table \ref{weights}. Although how to assign weights to achieve better performance is not studied in this paper, we maintain the percentage of weights on three operator class in Category and Cons configurations. But we modified some weights on the operators inside the same class, and we assigned different weights for different data sets.

\paragraph*{3.2 Simulated data sets}
We simulated two sets of sequence alignment on the same tree with 20 taxa that is shown in Figure \ref{simulateddatatree}. We used HKY model as substitution model with $\kappa = 2.4751$, and the base frequencies are $\pi = (0.2193 0.2268 0.3007 0.2531)$. In the uncorrelated relaxed clock model, the standard deviation of the branch rates (Ucldstdev) is 0.1803. The models and prior distributions are the same as is described in Figure \ref{modelvalidation}.

\paragraph*{3.3 Efficiency measured by ESS per hour}

Since we compare the efficiency based on ESS per hour using two configurations, i.e. Category and Cons, the ratio of ESS per hour is calculated by a random simulation in the two configurations, as is shown in Figure \ref{eff_comp}. Then Table \ref{eff_comp1} lists the average running time and ESS of particular parameters in the simulations using different data sets. Here, we present the detailed running time and ESS of the simulations, which can be seen in Figure \ref{anolistimeess} to Figure \ref{simulatedtimeess}. Overall, we conclude that the proposed operators are able to provide better performance, because the figures suggest that Cons configuration requires less running time and have larger ESS for most parameters in most simulations. Especially, for those poorly estimated parameters in Category configuration, the improvement is more obvious. For data sets such as primates and simulated data with 500 sites, the running time is slightly larger in Cons configuration, but the ESS are much larger, which makes it acceptable to reduce the MCMC chain length and get the same performance.

\paragraph*{3.4 Efficiency measured by proposals}

The operators introduced in the paper utilise a random walk proposal for the new node time, which draws a random number from a uniform distribution and moves the node uniformly on the branch. However, others proposals, such as a Bactrian proposal \cite{Yang19307} and a Beta proposal \cite{betaproposal}, assign a specific distribution on the new node time so that it is more probable to move to a certain height on the branch, either far away from or close to its original position. This section applied Random walk proposal (the operators in this paper), Bactrian proposal and Beta proposal to the three data sets, and the results are compared to those using Category configuration.

The comparisons are shown in Figure \ref{anolisproposal}, Figure \ref{rsv2proposal} and Figure \ref{shankarappaproposal}. It is indicated that Beta proposal achieved worst performance in the three analysed data sets. The performance of the Constant Distance operator (Random walk) and Bactrian proposal varies depending on different data sets. However, these two proposal methods are both more efficient that the Category configuration. Therefore, it still needs further investigation to demonstrate the effectiveness of different proposals.

\subsection*{4. UcldstdevScaleOperator: a scale operator on standard deviation}\label{UcldstdevScaler}

It should be noted that the proposed ConstantDistance operator parameterises branch rates as continuous random variables, instead of discrete rate categories as is used in current BEAST2 settings. In uncorrelated relaxed clock model, branch rates are assumed to have a lognormal prior distribution, where the real mean is fixed to 1 and the standard deviation (denoted by Ucldstdev) is usually sampled with a hyper prior such as gamma($\alpha = 0.5396, \beta = 0.3819$). When a new Ucldstdev is proposed in one state during MCMC sampling by normal operators, the probability of all rates change as well under the new log normal distribution. Therefore, the authors implemented a separate operator working on Ucldstdev, which is able to solve this problem properly. 

The first step is to propose a new Ucldstdev by a scale operation, which multiplies  current Ucldstdev by a random factor, as is shown in equation (\ref{ucldstdev_multiplier}).
\begin{equation}\label{ucldstdev_multiplier}
Ucldstdev' = Ucldstdev \times {\rm{scale}}
\end{equation}
where ${\rm{scale = Factor  +  }}\left[ {\xi  \times {\rm{(}}\frac{1}{{{\rm{Factor}}}}{\rm{  -  Factor)}}} \right]$ and  $\xi$ is a random variable from a $Uniform(0,1)$,  ${\rm{Factor}}$ is a user-defined parameter to specify how bold the proposal is.

Secondly, all the branch rates are proposed based on the new $Ucldstdev'$, given the probability of original $Ucldstdev$, which is calculated using equation (\ref{ucldstdev_rates}).
\begin{equation}\label{ucldstdev_rates}
r_{i}' = icd{f_{stdev'}}\left[ {cd{f_{stdev}}({r_i})} \right]
\end{equation}
where the notations $cdf(\cdot)$ and $icdf(\cdot)$ represent the cumulative and inverse cumulative density function of log normal distribution.

Finally, it is important to return the corrected hastings ratio, since the proposal is associated with one random variable, $Ucldstdev$ and $(2n-1)$ branch rates. As is shown in equation (\ref{ucldstdev_hr}), the ratio includes the scale operation and rates changing under the same probability.
\begin{equation}\label{ucldstdev_hr}
{{\mathbf{J}}_{Ucldstdev}} = \frac{1}{{scale}} \times \prod\limits_{i = 1}^{2n - 1} {\frac{{\partial icd{f_{Ucldstdev'}}[cd{f_{Ucldstdev}}({r_i})]}}{{\partial {r_i}}}}
\end{equation}


\end{backmatter}
\end{document}
